File: C++/n-ary_tree_preorder_traversal.cpp
#include "./n-ary_tree_preorder_traversal.h"
#include <stack>
#include <iostream>

using namespace std;

vector<int> Solution::preorder(Node* root) {
    stack<Node*> temp;
    vector<int> output;
    temp.push(root);
    while (temp.size() > 0) {
        Node* this_node = temp.top();
        temp.pop();
        if (this_node->children.size() > 0) {
            for (vector<Node*>::iterator iter = (this_node->children.end())-1; iter != this_node->children.begin(); --iter) {
                cout << iter->val << "\n";
                temp.push(*iter);
            }
        }
        output.push_back(this_node->val);
    }
    
    return output;
}


int main() {
    Node tree;
    tree.val = 1;
    tree.children.push_back(new Node());
    tree.children.back()->val = 3;
    tree.children.back()->children.push_back(new Node());
    tree.children.back()->children.back()->val = 5;
    tree.children.back()->children.push_back(new Node());
    tree.children.back()->children.back()->val = 6;
    tree.children.push_back(new Node());
    tree.children.back()->val = 2;
    tree.children.push_back(new Node());
    tree.children.back()->val = 4;

    Solution* sol = new Solution();
    vector<int> output = sol->preorder(&tree);
    Utility::printVector(output);
    return 0;
}

File: Python2/n-ary_tree_preorder_traversal.py
class Node(object):
    def __init__(self, val, children):
        self.val = val
        self.children = children


class Solution(object):
    def preorder(self, root):
        if not root:
            return []
        stack, output = [root], []
        while stack:
            this_node = stack.pop()
            output.append(this_node.val)
            if this_node.children:
                stack.extend(this_node.children[::-1])
        # O(N). Because we append the child nodes for each visit
        return output


if __name__ == '__main__':
    tree = \
        Node(1,
             [Node(3, [Node(5, []), Node(6, [])]), Node(2, []), Node(4, [])])
    sol = Solution()
    print sol.preorder(tree)
